Hotfix: Change http status code for client 2.4.2

micbar opened the pull request:
Hotfix for #144 

---

## Problem
The 2.4.2 desktop client goes into an infinite loop when he gets a "invalid_grant" while refreshing his access code with response code 400. This puts very high loads on the webserver.

## Hotfix
If he would get the same response with http response code 200, he would stop his infinite loop and open the webUI to re authorize.

## Changes
I changed the http response code to 200 after checking for the 2.4.2 client in the request header.

All other clients get the 400 response code, as the Oauth Specs demand it. https://tools.ietf.org/html/rfc6749#section-5.2

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/owncloud/oauth2?pullRequest=145) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/owncloud/oauth2?pullRequest=145) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/owncloud/oauth2?pullRequest=145) it.</sub>

patrickjahns commented on lib/Controller/OAuthApiController.php:
```diff
@@ -156,21 +156,27 @@ public function generateToken($grant_type, $code = null,
+					$statusCode = Http::STATUS_OK;
+				}
+				else {
+					$statusCode = Http::STATUS_BAD_REQUEST;
```
How about moving this before the if condition? 
```
$statusCode = Http::STATUS_BAD_REQUEST;
if (preg_match('/\bmirall\b.+2\.4\.2/i', $this->request->getHeader('User-Agent'))) {
	$statusCode = Http::STATUS_OK;
}
```

This eases the flow of reading - and you could add a comment on top outlining that this is a fix

micbar replied:
Good point!

micbar replied:
can you tell my, how to set a http user agent header in the unit test case?

micbar commented:
Output from travis/ci
```
$ php --version
PHP 7.4.0-dev (cli) (built: Aug  9 2018 06:03:59) ( ZTS )
Copyright (c) 1997-2018 The PHP Group
Zend Engine v3.4.0-dev, Copyright (c) 1998-2018 Zend Technologies
    with Zend OPcache v7.4.0-dev, Copyright (c) 1999-2018, by Zend Technologies
$ composer --version
Composer version 1.7.1 2018-08-07 09:39:23
before_install.1
0.00s$ export DISPLAY=:99.0
before_install.2
0.01s$ sh -e /etc/init.d/xvfb start
Starting virtual X frame buffer: Xvfb.
before_install.3
4.39s$ nvm install 8
Downloading and installing node v8.11.3...
Downloading https://nodejs.org/dist/v8.11.3/node-v8.11.3-linux-x64.tar.xz...
Computing checksum with sha256sum
Checksums matched!
Now using node v8.11.3 (npm v5.6.0)
before_install.4
9.06s$ npm install -g npm@latest
/home/travis/.nvm/versions/node/v8.11.3/bin/npm -> /home/travis/.nvm/versions/node/v8.11.3/lib/node_modules/npm/bin/npm-cli.js
/home/travis/.nvm/versions/node/v8.11.3/bin/npx -> /home/travis/.nvm/versions/node/v8.11.3/lib/node_modules/npm/bin/npx-cli.js
+ npm@6.3.0
added 283 packages, removed 363 packages and updated 41 packages in 8.647s
before_install.5
0.01s$ make
make: Nothing to be done for `all'.
before_install.6
0.06s$ make dist
rm -Rf /home/travis/build/owncloud/oauth2/build/dist
rm -Rf /home/travis/build/owncloud/oauth2/build/dist/oauth2; mkdir -p /home/travis/build/owncloud/oauth2/build/dist/oauth2
cp -R appinfo css img js l10n lib templates COPYING README.md CHANGELOG.md AUTHORS.md /home/travis/build/owncloud/oauth2/build/dist/oauth2
Skipping signing, either no key and certificate found in /home/travis/.owncloud/certificates/oauth2.key and /home/travis/.owncloud/certificates/oauth2.crt or occ can not be found at /home/travis/build/owncloud/oauth2/../../occ
tar -czf /home/travis/build/owncloud/oauth2/build/dist/oauth2.tar.gz -C /home/travis/build/owncloud/oauth2/build/dist oauth2
tar -cjf /home/travis/build/owncloud/oauth2/build/dist/oauth2.tar.bz2 -C /home/travis/build/owncloud/oauth2/build/dist oauth2
before_install.7
92.88s$ bash tests/.travis/before_install.sh $APP_NAME $CORE_BRANCH $DB
Work directory: /home/travis/build/owncloud/oauth2
Database: sqlite
Cloning into 'core'...
remote: Counting objects: 7080, done.
remote: Compressing objects: 100% (6188/6188), done.
remote: Total 7080 (delta 1346), reused 2544 (delta 662), pack-reused 0
Receiving objects: 100% (7080/7080), 10.92 MiB | 13.62 MiB/s, done.
Resolving deltas: 100% (1346/1346), done.
Building core
Note: You can type 'make help' for more targets
cd build && ./getcomposer.sh
php build/composer.phar install --dev
You are using the deprecated option "dev". Dev packages are installed by default now.
Loading composer repositories with package information
Installing dependencies (including require-dev) from lock file
Package operations: 121 installs, 0 updates, 0 removals
  - Installing bantu/ini-get-wrapper (v1.0.1): Downloading (100%)
  - Installing deepdiver/zipstreamer (1.1.1): Downloading (100%)
  - Installing deepdiver1975/tarstreamer (v0.1.0): Downloading (100%)
  - Installing doctrine/lexer (v1.0.1): Downloading (100%)
  - Installing doctrine/annotations (v1.4.0): Downloading (100%)
  - Installing doctrine/cache (v1.6.2): Downloading (100%)
  - Installing doctrine/collections (v1.4.0): Downloading (100%)
  - Installing doctrine/inflector (v1.1.0): Downloading (100%)
  - Installing doctrine/common (v2.7.3): Downloading (100%)
  - Installing doctrine/dbal (v2.5.13): Downloading (100%)
  - Installing guzzle/parser (v3.8.1): Downloading (100%)
  - Installing symfony/event-dispatcher (v3.4.14): Downloading (100%)
  - Installing guzzle/common (v3.8.1): Downloading (100%)
  - Installing guzzle/stream (v3.8.1): Downloading (100%)
  - Installing react/promise (v2.7.0): Downloading (100%)
  - Installing guzzlehttp/streams (3.0.0): Downloading (100%)
  - Installing guzzlehttp/ringphp (1.1.1): Downloading (100%)
  - Installing icewind/streams (0.5.2): Downloading (100%)
  - Installing interfasys/lognormalizer (v1.0): Downloading (100%)
  - Installing nikic/php-parser (v1.4.1): Downloading (100%)
  - Installing jeremeamia/superclosure (2.1.0): Downloading (100%)
  - Installing league/flysystem (1.0.45): Downloading (100%)
  - Installing lukasreschke/id3parser (v0.0.3): Downloading (100%)
  - Installing patchwork/jsqueeze (v2.0.5): Downloading (100%)
  - Installing patchwork/utf8 (v1.3.1): Downloading (100%)
  - Installing pear/pear_exception (v1.0.0): Downloading (100%)
  - Installing pear/console_getopt (v1.4.1): Downloading (100%)
  - Installing pear/pear-core-minimal (v1.10.3): Downloading (100%)
  - Installing pear/archive_tar (1.4.3): Downloading (100%)
  - Installing phpseclib/phpseclib (2.0.11): Downloading (100%)
  - Installing pimple/pimple (v3.0.2): Downloading (100%)
  - Installing punic/punic (3.1.0): Downloading (100%)
  - Installing guzzle/http (v3.8.1): Downloading (100%)
  - Installing rackspace/php-opencloud (v1.9.2): Downloading (100%)
  - Installing sabre/uri (1.2.1): Downloading (100%)
  - Installing sabre/xml (1.5.0): Downloading (100%)
  - Installing sabre/vobject (4.1.6): Downloading (100%)
  - Installing sabre/event (3.0.0): Downloading (100%)
  - Installing sabre/http (v4.2.4): Downloading (100%)
  - Installing psr/log (1.0.2): Downloading (100%)
  - Installing sabre/dav (3.2.2): Downloading (100%)
  - Installing swiftmailer/swiftmailer (v5.4.12): Downloading (100%)
  - Installing symfony/polyfill-mbstring (v1.9.0): Downloading (100%)
  - Installing symfony/debug (v3.4.14): Downloading (100%)
  - Installing symfony/console (v3.4.14): Downloading (100%)
  - Installing paragonie/random_compat (v2.0.17): Downloading (100%)
  - Installing symfony/polyfill-php70 (v1.9.0): Downloading (100%)
  - Installing symfony/process (v3.4.14): Downloading (100%)
  - Installing symfony/routing (v3.4.14): Downloading (100%)
  - Installing symfony/translation (v3.4.14): Downloading (100%)
  - Installing zendframework/zend-stdlib (3.2.0): Downloading (100%)
  - Installing psr/container (1.0.0): Downloading (100%)
  - Installing container-interop/container-interop (1.2.0): Downloading (100%)
  - Installing zendframework/zend-validator (2.10.2): Downloading (100%)
  - Installing zendframework/zend-servicemanager (3.3.2): Downloading (100%)
  - Installing zendframework/zend-filter (2.8.0): Downloading (100%)
  - Installing zendframework/zend-inputfilter (2.8.2): Downloading (100%)
  - Installing behat/gherkin (v4.5.1): Downloading (100%)
  - Installing symfony/polyfill-ctype (v1.9.0): Downloading (100%)
  - Installing symfony/dom-crawler (v2.8.44): Downloading (100%)
  - Installing symfony/css-selector (v2.8.44): Downloading (100%)
  - Installing symfony/browser-kit (v2.8.44): Downloading (100%)
  - Installing guzzlehttp/guzzle (5.3.3): Downloading (100%)
  - Installing fabpot/goutte (v2.0.4): Downloading (100%)
  - Installing behat/mink (v1.7.1): Downloading (100%)
  - Installing behat/mink-browserkit-driver (1.3.3): Downloading (100%)
  - Installing behat/mink-goutte-driver (v1.2.1): Downloading (100%)
  - Installing instaclick/php-webdriver (1.4.5): Downloading (100%)
  - Installing behat/mink-selenium2-driver (dev-master f4efaf5): Cloning f4efaf52fe from cache
  - Installing behat/transliterator (v1.2.0): Downloading (100%)
  - Installing composer/semver (1.4.2): Downloading (100%)
  - Installing composer/xdebug-handler (1.1.0): Downloading (100%)
  - Installing jakub-onderka/php-console-color (0.1): Downloading (100%)
  - Installing jakub-onderka/php-console-highlighter (v0.3.2): Downloading (100%)
  - Installing jakub-onderka/php-parallel-lint (v1.0.0): Downloading (100%)
  - Installing symfony/yaml (v3.4.14): Downloading (100%)
  - Installing symfony/dependency-injection (v3.4.14): Downloading (100%)
  - Installing symfony/filesystem (v3.4.14): Downloading (100%)
  - Installing symfony/config (v3.4.14): Downloading (100%)
  - Installing symfony/class-loader (v3.4.14): Downloading (100%)
  - Installing behat/behat (v3.4.3): Downloading (100%)
  - Installing jarnaiz/behat-junit-formatter (v1.3.2): Downloading (100%)
  - Installing mikey179/vfsstream (v1.6.5): Downloading (100%)
  - Installing symfony/stopwatch (v3.4.14): Downloading (100%)
  - Installing symfony/polyfill-php72 (v1.9.0): Downloading (100%)
  - Installing symfony/options-resolver (v3.4.14): Downloading (100%)
  - Installing symfony/finder (v3.4.14): Downloading (100%)
  - Installing php-cs-fixer/diff (v1.3.0): Downloading (100%)
  - Installing friendsofphp/php-cs-fixer (v2.12.2): Downloading (100%)
  - Installing owncloud/coding-standard (1.0.1): Downloading (100%)
  - Installing webmozart/assert (1.3.0): Downloading (100%)
  - Installing phpdocumentor/reflection-common (1.0.1): Downloading (100%)
  - Installing phpdocumentor/type-resolver (0.4.0): Downloading (100%)
  - Installing phpdocumentor/reflection-docblock (3.3.2): Downloading (100%)
  - Installing phpunit/php-token-stream (1.4.12): Downloading (100%)
  - Installing sebastian/version (2.0.1): Downloading (100%)
  - Installing sebastian/resource-operations (1.0.0): Downloading (100%)
  - Installing sebastian/recursion-context (2.0.0): Downloading (100%)
  - Installing sebastian/object-enumerator (2.0.1): Downloading (100%)
  - Installing sebastian/global-state (1.1.1): Downloading (100%)
  - Installing sebastian/exporter (2.0.0): Downloading (100%)
  - Installing sebastian/environment (2.0.0): Downloading (100%)
  - Installing sebastian/diff (1.4.3): Downloading (100%)
  - Installing sebastian/comparator (1.2.4): Downloading (100%)
  - Installing phpunit/php-text-template (1.2.1): Downloading (100%)
  - Installing doctrine/instantiator (1.0.5): Downloading (100%)
  - Installing phpunit/phpunit-mock-objects (3.4.4): Downloading (100%)
  - Installing phpunit/php-timer (1.0.9): Downloading (100%)
  - Installing phpunit/php-file-iterator (1.4.5): Downloading (100%)
  - Installing sebastian/code-unit-reverse-lookup (1.0.1): Downloading (100%)
  - Installing phpunit/php-code-coverage (4.0.8): Downloading (100%)
  - Installing phpspec/prophecy (1.8.0): Downloading (100%)
  - Installing myclabs/deep-copy (1.7.0): Downloading (100%)
  - Installing phpunit/phpunit (5.7.27): Downloading (100%)
  - Installing rdx/behat-variables (1.2): Downloading (100%)
  - Installing zendframework/zend-eventmanager (3.2.1): Downloading (100%)
  - Installing zendframework/zend-code (2.6.3): Downloading (100%)
  - Installing ocramius/proxy-manager (1.0.2): Downloading (100%)
  - Installing behat/mink-extension (2.3.1): Downloading (100%)
  - Installing sensiolabs/behat-page-object-extension (v2.1.0): Downloading (100%)
symfony/event-dispatcher suggests installing symfony/http-kernel
league/flysystem suggests installing league/flysystem-aws-s3-v2 (Allows you to use S3 storage with AWS SDK v2)
league/flysystem suggests installing league/flysystem-aws-s3-v3 (Allows you to use S3 storage with AWS SDK v3)
league/flysystem suggests installing league/flysystem-azure (Allows you to use Windows Azure Blob storage)
league/flysystem suggests installing league/flysystem-cached-adapter (Flysystem adapter decorator for metadata caching)
league/flysystem suggests installing league/flysystem-eventable-filesystem (Allows you to use EventableFilesystem)
league/flysystem suggests installing league/flysystem-rackspace (Allows you to use Rackspace Cloud Files)
league/flysystem suggests installing league/flysystem-sftp (Allows you to use SFTP server storage via phpseclib)
league/flysystem suggests installing league/flysystem-webdav (Allows you to use WebDAV storage)
league/flysystem suggests installing league/flysystem-ziparchive (Allows you to use ZipArchive adapter)
league/flysystem suggests installing spatie/flysystem-dropbox (Allows you to use Dropbox storage)
league/flysystem suggests installing srmklive/flysystem-dropbox-v2 (Allows you to use Dropbox storage for PHP 5 applications)
patchwork/utf8 suggests installing ext-wfio (Use WFIO for UTF-8 filesystem access on Windows)
pear/archive_tar suggests installing ext-xz (lzma2 compression support.)
phpseclib/phpseclib suggests installing ext-libsodium (SSH2/SFTP can make use of some algorithms provided by the libsodium-php extension.)
phpseclib/phpseclib suggests installing ext-mcrypt (Install the Mcrypt extension in order to speed up a few other cryptographic operations.)
sabre/vobject suggests installing hoa/bench (If you would like to run the benchmark scripts)
symfony/console suggests installing psr/log-implementation (For using the console logger)
symfony/console suggests installing symfony/lock
paragonie/random_compat suggests installing ext-libsodium (Provides a modern crypto API that can be used to generate random bytes.)
symfony/routing suggests installing symfony/expression-language (For using expression matching)
symfony/routing suggests installing symfony/http-foundation (For using a Symfony Request object)
symfony/translation suggests installing psr/log-implementation (To use logging capability in translator)
zendframework/zend-validator suggests installing zendframework/zend-db (Zend\Db component, required by the (No)RecordExists validator)
zendframework/zend-validator suggests installing zendframework/zend-i18n (Zend\I18n component to allow translation of validation error messages)
zendframework/zend-validator suggests installing zendframework/zend-i18n-resources (Translations of validator messages)
zendframework/zend-validator suggests installing zendframework/zend-math (Zend\Math component, required by the Csrf validator)
zendframework/zend-validator suggests installing zendframework/zend-session (Zend\Session component, ^2.8; required by the Csrf validator)
zendframework/zend-validator suggests installing zendframework/zend-uri (Zend\Uri component, required by the Uri and Sitemap\Loc validators)
zendframework/zend-filter suggests installing zendframework/zend-crypt (Zend\Crypt component, for encryption filters)
zendframework/zend-filter suggests installing zendframework/zend-i18n (Zend\I18n component for filters depending on i18n functionality)
zendframework/zend-filter suggests installing zendframework/zend-uri (Zend\Uri component, for the UriNormalize filter)
behat/mink suggests installing behat/mink-zombie-driver (fast and JS-enabled headless driver for any app (requires node.js))
symfony/dependency-injection suggests installing symfony/expression-language (For using expressions in service container configuration)
symfony/dependency-injection suggests installing symfony/proxy-manager-bridge (Generate service proxies to lazy load them)
symfony/class-loader suggests installing symfony/polyfill-apcu (For using ApcClassLoader on HHVM)
behat/behat suggests installing behat/symfony2-extension (for integration with Symfony2 web framework)
behat/behat suggests installing behat/yii-extension (for integration with Yii web framework)
friendsofphp/php-cs-fixer suggests installing php-cs-fixer/phpunit-constraint-isidenticalstring (For IsIdenticalString constraint.)
friendsofphp/php-cs-fixer suggests installing php-cs-fixer/phpunit-constraint-xmlmatchesxsd (For XmlMatchesXsd constraint.)
sebastian/global-state suggests installing ext-uopz (*)
phpunit/php-code-coverage suggests installing ext-xdebug (^2.5.1)
phpunit/phpunit suggests installing ext-xdebug (*)
phpunit/phpunit suggests installing phpunit/php-invoker (~1.1)
ocramius/proxy-manager suggests installing ocramius/generated-hydrator (To have very fast object to array to object conversion for ghost objects)
ocramius/proxy-manager suggests installing zendframework/zend-json (To have the JsonRpc adapter (Remote Object feature))
ocramius/proxy-manager suggests installing zendframework/zend-soap (To have the Soap adapter (Remote Object feature))
ocramius/proxy-manager suggests installing zendframework/zend-xmlrpc (To have the XmlRpc adapter (Remote Object feature))
sensiolabs/behat-page-object-extension suggests installing bossa/phpspec2-expect (Allows to use PHPSpec2 matchers in Behat context files)
Package guzzle/parser is abandoned, you should avoid using it. Use guzzle/guzzle instead.
Package guzzle/common is abandoned, you should avoid using it. Use guzzle/guzzle instead.
Package guzzle/stream is abandoned, you should avoid using it. Use guzzle/guzzle instead.
Package guzzle/http is abandoned, you should avoid using it. Use guzzle/guzzle instead.
Generating optimized autoload files
/home/travis/.nvm/versions/node/v8.11.3/bin/npm install --prefix build
> uws@9.14.0 install /home/travis/build/owncloud/core/build/node_modules/uws
> node-gyp rebuild > build_log.txt 2>&1 || exit 0
> phantomjs-prebuilt@2.1.16 install /home/travis/build/owncloud/core/build/node_modules/phantomjs-prebuilt
> node install.js
Considering PhantomJS found at /usr/local/phantomjs/bin/phantomjs
Found PhantomJS at /usr/local/phantomjs/bin/phantomjs ...verifying
Writing location.js file
PhantomJS is already installed on PATH at /usr/local/phantomjs/bin/phantomjs
npm WARN optional SKIPPING OPTIONAL DEPENDENCY: fsevents@1.2.4 (node_modules/fsevents):
npm WARN notsup SKIPPING OPTIONAL DEPENDENCY: Unsupported platform for fsevents@1.2.4: wanted {"os":"darwin","arch":"any"} (current: {"os":"linux","arch":"x64"})
added 558 packages from 1127 contributors and audited 2329 packages in 24.95s
found 11 vulnerabilities (1 low, 10 moderate)
  run `npm audit fix` to fix them, or `npm audit` for details
touch build/node_modules
build/node_modules/bower/bin/bower install
bower invalid-meta  for:/home/travis/build/owncloud/core/bower.json
bower invalid-meta  The "name" is recommended to be lowercase, can contain digits, dots, dashes
bower not-cached    https://github.com/evert/davclient.js.git#0.1.2
bower resolve       https://github.com/evert/davclient.js.git#0.1.2
bower not-cached    https://github.com/jakearchibald/es6-promise.git#2.3.0
bower resolve       https://github.com/jakearchibald/es6-promise.git#2.3.0
bower not-cached    https://github.com/blueimp/JavaScript-MD5.git#1.0.1
bower resolve       https://github.com/blueimp/JavaScript-MD5.git#1.0.1
bower not-cached    https://github.com/twbs/bootstrap.git#3.3.6
bower resolve       https://github.com/twbs/bootstrap.git#3.3.6
bower not-cached    https://github.com/ivaynberg/select2.git#3.5.4
bower resolve       https://github.com/ivaynberg/select2.git#3.5.4
bower not-cached    https://github.com/jakiestfu/Snap.js.git#2.0.0-rc1
bower resolve       https://github.com/jakiestfu/Snap.js.git#2.0.0-rc1
bower not-cached    https://github.com/davidchambers/Base64.js.git#0.3.0
bower resolve       https://github.com/davidchambers/Base64.js.git#0.3.0
bower not-cached    https://github.com/moment/moment.git#2.10.6
bower resolve       https://github.com/moment/moment.git#2.10.6
bower not-cached    https://github.com/Graffino/Browser-Update.git#^2.0.1
bower resolve       https://github.com/Graffino/Browser-Update.git#^2.0.1
bower not-cached    https://github.com/dropbox/zxcvbn.git#*
bower resolve       https://github.com/dropbox/zxcvbn.git#*
bower not-cached    https://github.com/components/handlebars.js.git#4.0.5
bower resolve       https://github.com/components/handlebars.js.git#4.0.5
bower not-cached    https://github.com/jashkenas/underscore.git#1.8.3
bower resolve       https://github.com/jashkenas/underscore.git#1.8.3
bower not-cached    https://github.com/MorrisJobke/strengthify.git#0.5.0
bower resolve       https://github.com/MorrisJobke/strengthify.git#0.5.0
bower not-cached    https://github.com/tapmodo/Jcrop.git#0.9.12
bower resolve       https://github.com/tapmodo/Jcrop.git#0.9.12
bower not-cached    https://github.com/zenorocha/clipboard.js.git#1.5.12
bower resolve       https://github.com/zenorocha/clipboard.js.git#1.5.12
bower not-cached    https://github.com/showdownjs/showdown.git#^1.7.1
bower resolve       https://github.com/showdownjs/showdown.git#^1.7.1
bower not-cached    https://github.com/components/jqueryui.git#1.10.0
bower resolve       https://github.com/components/jqueryui.git#1.10.0
bower not-cached    https://github.com/appleboy/jquery-migrate.git#1.4.0
bower resolve       https://github.com/appleboy/jquery-migrate.git#1.4.0
bower not-cached    https://github.com/HenningM/jstimezonedetect.git#1.0.5
bower resolve       https://github.com/HenningM/jstimezonedetect.git#1.0.5
bower not-cached    https://github.com/ded/bowser.git#^1.6.0
bower resolve       https://github.com/ded/bowser.git#^1.6.0
bower not-cached    https://github.com/jquery/jquery-dist.git#2.1.4
bower resolve       https://github.com/jquery/jquery-dist.git#2.1.4
bower not-cached    https://github.com/jashkenas/backbone.git#1.2.3
bower resolve       https://github.com/jashkenas/backbone.git#1.2.3
bower download      https://github.com/evert/davclient.js/archive/0.1.2.tar.gz
bower download      https://github.com/jakearchibald/es6-promise/archive/2.3.0.tar.gz
bower download      https://github.com/blueimp/JavaScript-MD5/archive/1.0.1.tar.gz
bower download      https://github.com/ivaynberg/select2/archive/3.5.4.tar.gz
bower download      https://github.com/twbs/bootstrap/archive/v3.3.6.tar.gz
bower download      https://github.com/jakiestfu/Snap.js/archive/v2.0.0-rc1.tar.gz
bower download      https://github.com/davidchambers/Base64.js/archive/0.3.0.tar.gz
bower extract       blueimp-md5#1.0.1 archive.tar.gz
bower download      https://github.com/moment/moment/archive/2.10.6.tar.gz
bower download      https://github.com/Graffino/Browser-Update/archive/v2.0.2.tar.gz
bower extract       davclient.js#0.1.2 archive.tar.gz
bower extract       es6-promise#2.3.0 archive.tar.gz
bower download      https://github.com/dropbox/zxcvbn/archive/v4.4.2.tar.gz
bower download      https://github.com/components/handlebars.js/archive/v4.0.5.tar.gz
bower resolved      https://github.com/blueimp/JavaScript-MD5.git#1.0.1
bower extract       snapjs#2.0.0-rc1 archive.tar.gz
bower extract       base64#0.3.0 archive.tar.gz
bower resolved      https://github.com/evert/davclient.js.git#0.1.2
bower download      https://github.com/jashkenas/underscore/archive/1.8.3.tar.gz
bower download      https://github.com/MorrisJobke/strengthify/archive/0.5.0.tar.gz
bower extract       browser-update#^2.0.1 archive.tar.gz
bower download      https://github.com/tapmodo/Jcrop/archive/v0.9.12.tar.gz
bower extract       select2#3.5.4 archive.tar.gz
bower download      https://github.com/showdownjs/showdown/archive/1.8.6.tar.gz
bower resolved      https://github.com/davidchambers/Base64.js.git#0.3.0
bower extract       moment#2.10.6 archive.tar.gz
bower extract       handlebars#4.0.5 archive.tar.gz
bower download      https://github.com/zenorocha/clipboard.js/archive/v1.5.12.tar.gz
bower extract       bootstrap#3.3.6 archive.tar.gz
bower resolved      https://github.com/Graffino/Browser-Update.git#2.0.2
bower extract       strengthify#0.5.0 archive.tar.gz
bower download      https://github.com/components/jqueryui/archive/1.10.0.tar.gz
bower download      https://github.com/appleboy/jquery-migrate/archive/1.4.0.tar.gz
bower extract       underscore#1.8.3 archive.tar.gz
bower download      https://github.com/HenningM/jstimezonedetect/archive/v1.0.5.tar.gz
bower resolved      https://github.com/MorrisJobke/strengthify.git#0.5.0
bower invalid-meta  for:/tmp/travis/bower/4ac41f654a41571ecbe1f552e76c4293-5186-iHgjOU/bower.json
bower invalid-meta  The "name" is recommended to be lowercase, can contain digits, dots, dashes
bower extract       jcrop#0.9.12 archive.tar.gz
bower resolved      https://github.com/jakiestfu/Snap.js.git#2.0.0-rc1
bower extract       zxcvbn#* archive.tar.gz
bower extract       showdown#^1.7.1 archive.tar.gz
bower extract       clipboard#1.5.12 archive.tar.gz
bower download      https://github.com/ded/bowser/archive/1.9.4.tar.gz
bower resolved      https://github.com/components/handlebars.js.git#4.0.5
bower download      https://github.com/jquery/jquery-dist/archive/2.1.4.tar.gz
bower resolved      https://github.com/jakearchibald/es6-promise.git#2.3.0
bower resolved      https://github.com/tapmodo/Jcrop.git#0.9.12
bower invalid-meta  for:/tmp/travis/bower/35368a19f307e4af02d0df055846840d-5186-QL0nas/bower.json
bower invalid-meta  The "main" field cannot contain font, image, audio, or video files
bower invalid-meta  The "main" field cannot contain font, image, audio, or video files
bower invalid-meta  The "main" field cannot contain font, image, audio, or video files
bower invalid-meta  The "main" field has to contain only 1 file per filetype; found multiple .png files: ["select2.png","select2x2.png"]
bower resolved      https://github.com/ivaynberg/select2.git#3.5.4
bower extract       jquery-migrate#1.4.0 archive.tar.gz
bower resolved      https://github.com/zenorocha/clipboard.js.git#1.5.12
bower extract       jsTimezoneDetect#1.0.5 archive.tar.gz
bower resolved      https://github.com/appleboy/jquery-migrate.git#1.4.0
bower extract       jquery-ui#1.10.0 archive.tar.gz
bower resolved      https://github.com/jashkenas/underscore.git#1.8.3
bower invalid-meta  for:/tmp/travis/bower/8c85a860e5e0cea46e1e41b37af1790b-5186-TMvskO/bower.json
bower invalid-meta  The "name" is recommended to be lowercase, can contain digits, dots, dashes
bower invalid-meta  The "main" field cannot contain minified files
bower download      https://github.com/jashkenas/backbone/archive/1.2.3.tar.gz
bower resolved      https://github.com/HenningM/jstimezonedetect.git#1.0.5
bower resolved      https://github.com/dropbox/zxcvbn.git#4.4.2
bower extract       jquery#2.1.4 archive.tar.gz
bower resolved      https://github.com/showdownjs/showdown.git#1.8.6
bower extract       bowser#^1.6.0 archive.tar.gz
bower resolved      https://github.com/ded/bowser.git#1.9.4
bower extract       backbone#1.2.3 archive.tar.gz
bower resolved      https://github.com/jquery/jquery-dist.git#2.1.4
bower resolved      https://github.com/jashkenas/backbone.git#1.2.3
bower resolved      https://github.com/twbs/bootstrap.git#3.3.6
bower resolved      https://github.com/moment/moment.git#2.10.6
bower resolved      https://github.com/components/jqueryui.git#1.10.0
bower install       blueimp-md5#1.0.1
bower install       davclient.js#0.1.2
bower install       base64#0.3.0
bower install       browser-update#2.0.2
bower install       strengthify#0.5.0
bower install       snapjs#2.0.0-rc1
bower install       handlebars#4.0.5
bower install       es6-promise#2.3.0
bower install       jcrop#0.9.12
bower install       select2#3.5.4
bower install       clipboard#1.5.12
bower install       jquery-migrate#1.4.0
bower install       underscore#1.8.3
bower install       jsTimezoneDetect#1.0.5
bower install       zxcvbn#4.4.2
bower install       showdown#1.8.6
bower install       bowser#1.9.4
bower install       jquery#2.1.4
bower install       backbone#1.2.3
bower install       bootstrap#3.3.6
bower install       moment#2.10.6
bower install       jquery-ui#1.10.0
blueimp-md5#1.0.1 core/vendor/blueimp-md5
davclient.js#0.1.2 core/vendor/davclient.js
base64#0.3.0 core/vendor/base64
browser-update#2.0.2 core/vendor/browser-update
strengthify#0.5.0 core/vendor/strengthify
snapjs#2.0.0-rc1 core/vendor/snapjs
handlebars#4.0.5 core/vendor/handlebars
es6-promise#2.3.0 core/vendor/es6-promise
jcrop#0.9.12 core/vendor/jcrop
select2#3.5.4 core/vendor/select2
└── jquery#2.1.4
clipboard#1.5.12 core/vendor/clipboard
jquery-migrate#1.4.0 core/vendor/jquery-migrate
underscore#1.8.3 core/vendor/underscore
jsTimezoneDetect#1.0.5 core/vendor/jsTimezoneDetect
zxcvbn#4.4.2 core/vendor/zxcvbn
showdown#1.8.6 core/vendor/showdown
bowser#1.9.4 core/vendor/bowser
jquery#2.1.4 core/vendor/jquery
backbone#1.2.3 core/vendor/backbone
└── underscore#1.8.3
bootstrap#3.3.6 core/vendor/bootstrap
└── jquery#2.1.4
moment#2.10.6 core/vendor/moment
jquery-ui#1.10.0 core/vendor/jquery-ui
└── jquery#2.1.4
File tests/.travis/custom.ini added to /home/travis/.phpenv/versions/nightly/etc/conf.d.
Using database oc_autotest
<?php
$AUTOCONFIG = array (
  'installed' => false,
  'dbtype' => 'oci',
  'dbtableprefix' => 'oc_',
  'adminlogin' => 'admin',
  'adminpass' => 'admin',
  'directory' => '/home/travis/build/owncloud/core/data-autotest',
  'dbuser' => 'autotest',
  'dbname' => 'XE',
  'dbhost' =>'',
  'dbpass' => 'owncloud',
  'loglevel' => 0,
);
Setup environment for sqlite testing ...
INDEX
This version of ownCloud is not compatible with PHP 7.3<br/>You are currently running PHP 7.4.0-dev.END INDEX
owncloud configuration:
<?php
$CONFIG = [
	'apps_paths' => [
		[
			'path'		=> OC::$SERVERROOT . '/apps',
			'url'		=> '/apps',
			'writable'	=> true,
		],
	],
	'default_language' => 'en',
];
if (\is_dir(OC::$SERVERROOT.'/apps2')) {
	$CONFIG['apps_paths'][] = [
		'path' => OC::$SERVERROOT . '/apps2',
		'url' => '/apps2',
		'writable' => false,
	];
}
if (\getenv("TC") === "selenium") {
	$CONFIG['skeletondirectory'] = OC::$SERVERROOT . '/tests/ui/skeleton';
}
data directory:
total 0
owncloud.log:
before_install.8
0.00s$ cd ../core
0.06s$ ./occ app:enable $APP_NAME
This version of ownCloud is not compatible with PHP 7.3
You are currently running PHP 7.4.0-dev.
The command "./occ app:enable $APP_NAME" failed and exited with 1 during .
Your build has been stopped.
```

phil-davis commented:
I restarted the first Travis job 974.1 for PHP 5.6 - that had hung for some reason and got the "no output for 10 minutes" message.

phil-davis commented:
Where did you see stuff about PHP 7.3 and PHP 7.4 ???

micbar commented:
I saw that at the travis build log.

phil-davis commented:
Something went crazy in Travis! PHP 7.3beta1 came out last week, and so I guess PHP 7.4.0-dev does exist. But goodness knows what happened for Travis to use those.

Anyway, CI passes.

DeepDiver1975 commented:
Will review soon.

DeepDiver1975 commented on tests/Unit/Controller/OAuthApiControllerTest.php:
```diff
@@ -358,4 +364,85 @@ public function testGenerateTokenWithRefreshToken() {
+		$_SERVER['PHP_AUTH_PW'] = null;
+
+		$result = $this->controller->generateToken('refresh_token', null, null, $this->refreshToken->getToken());
+		$this->assertTrue($result instanceof JSONResponse);
```
assertInstanceOf

DeepDiver1975 commented on tests/Unit/Controller/OAuthApiControllerTest.php:
```diff
@@ -358,4 +364,85 @@ public function testGenerateTokenWithRefreshToken() {
+		$this->assertNotEmpty($json->message_url);
+		$this->assertEquals($this->authorizationSuccessfulMessageUrl, $json->message_url);
+		$this->assertEquals(200, $result->getStatus());
+		$this->assertEquals(1, count($this->accessTokenMapper->findAll()));
```
assertCount

DeepDiver1975 commented on tests/Unit/Controller/OAuthApiControllerTest.php:
```diff
@@ -358,4 +364,85 @@ public function testGenerateTokenWithRefreshToken() {
+
+		$result = $this->controller->generateToken('refresh_token', null, null, $this->refreshToken->getToken());
+		$this->assertTrue($result instanceof JSONResponse);
+		$json = json_decode($result->render());
```
just use getData - this gives you the array
